synapse/handlers/federation.py

```diff
@@ -292,38 +295,42 @@ async def _maybe_backfill_inner(
-        #   forward event but not the backward extremity, e.g. in the case of
-        #   initial join of the server where we are allowed to see the join
-        #   event but not anything before it. This would require looking at the
-        #   state *before* the event, ignoring the special casing certain event
-        #   types have.
-
-        forward_event_ids = await self.store.get_successor_events(
-            (e.event_id for e in backwards_extremities)
-        )
 
-        extremities_events = await self.store.get_events(
-            forward_event_ids,
-            redact_behaviour=EventRedactBehaviour.AS_IS,
-            get_prev_content=False,
-        )
+        found_filtered_extremity = False
+        for bp in backwards_extremities:
+            # Given we don't have the extremity events themselves, we
+            # need to actually check the events that reference them - their "successor"
+            # events.
+            #
+            # TODO: Correctly handle the case where we are allowed to see the
+            #   successor event but not the backward extremity, e.g. in the case of
+            #   initial join of the server where we are allowed to see the join
+            #   event but not anything before it. This would require looking at the
+            #   state *before* the event, ignoring the special casing certain event
+            #   types have.
+
+            forward_event_ids = await self.store.get_successor_events([bp.event_id])
+
+            extremities_events = await self.store.get_events(
+                forward_event_ids,
+                redact_behaviour=EventRedactBehaviour.AS_IS,
+                get_prev_content=False,
+            )
 
-        # We set `check_history_visibility_only` as we might otherwise get false
-        # positives from users having been erased.
-        filtered_extremities = await filter_events_for_server(
-            self.storage,
-            self.server_name,
-            list(extremities_events.values()),
-            redact=False,
-            check_history_visibility_only=True,
-        )
-        logger.debug(
-            "_maybe_backfill_inner: filtered_extremities %s", filtered_extremities
-        )
+            # We set `check_history_visibility_only` as we might otherwise get false
+            # positives from users having been erased.
+            filtered_extremities = await filter_events_for_server(
+                self.storage,
+                self.server_name,
+                list(extremities_events.values()),
+                redact=False,
+                check_history_visibility_only=True,
+            )
+            if filtered_extremities:
+                found_filtered_extremity = True
+                break
 
-        if not filtered_extremities and not insertion_events_to_be_backfilled:
+        if not found_filtered_extremity and not insertion_events_to_be_backfilled:
             return False
```

Comment:
To check: the only reason to compute `found_filtered_extremity` is to possibly allow us to return early here?

Reply:
Yes, though it's worth noting that https://github.com/matrix-org/synapse/pull/12522/commits/1f0a63e6b57ca94775408c3b3ad81ff0df3dbceb changes this.